fsck/fsck.c

```diff
@@ -677,6 +677,34 @@ static int check_inode(struct exfat_de_iter *iter, struct exfat_inode *node)
 	return valid ? ret : -EINVAL;
 }
 
+static int handle_duplicated_filename(struct exfat_de_iter *iter,
+		struct exfat_inode *inode)
+{
+	int ret;
+	char filename[PATH_MAX + 1] = {0};
+	struct exfat_lookup_filter filter;
+
+	ret = exfat_utf16_dec(inode->name, NAME_BUFFER_SIZE, filename,
+			PATH_MAX);
+	if (ret < 0) {
+		exfat_err("failed to decode filename\n");
+		return ret;
+	}
+
+	ret = exfat_lookup_file(iter->exfat, iter->parent, filename, &filter);
```

Comment:
exfat_lookup_file encodes the filename back to utf16 again. How about introducing a new exfat_lookup_fiie which accepts utf16.

Reply:
And we don't need to search the filename until the end of a directory.

Reply:
> exfat_lookup_file encodes the filename back to utf16 again. How about introducing a new exfat_lookup_fiie which accepts utf16.

It's worth considering.

> And we don't need to search the filename until the end of a directory.

Because the file name exists, exfat_lookup_file() will exit when the file name is found, and will not search to the end of the directory.

Reply:
> > And we don't need to search the filename until the end of a directory.
> 
> Because the file name exists, exfat_lookup_file() will exit when the file name is found, and will not search to the end of the directory.

It can not guarantee the existence, so it is better to search directory entries from the first to the entry specified by the filename.

Reply:
Do you guarantee that there is an another duplicated filename when a same checksum exists? If not, exfat_lookup_file() always finds out  itself if there is no the duplicated filename.

Reply:
ping?

Reply:
Your are right. I also missed that you checked the device offset.